File: backend/routes/users/controller.js
const { generateToken } = require('../../utils/jwt');
const bcrypt = require('bcrypt');
const prisma = require('../../utils/prisma');

exports.createNewAccount = async (req, res) => {
    const { name, email, password } = req.body;

    // Validate required fields
    if (!name || !email || !password) {
        return res.status(400).json({ 
            error: 'Missing required fields: name, surname, email, password' 
        });
    }
    // Check if user already exists
    const existingUser = await prisma.users.count({
        where: {
            email: email
        }
    });
    if (existingUser > 0) {
        return res.status(400).json({ 
            error: 'User already exists with this email' 
        });
    }
    // Create new user account
    const hashedPassword = await bcrypt.hash(password, 10); // 10 is the salt rounds
    await prisma.users.create({
        data: {
            name: name,
            email: email,
            password: hashedPassword,
            role: 0 // Default role
        }
    });
    // Get user ID and role
    const user = await prisma.users.findUnique({
        where: {
            email: email
        }
    });

    // create JWT token
    const token = generateToken({
        userId: user.id,
        name: user.name,
        email: user.email,
        role: user.role
    });

    res.status(200).json({
        message: 'Account created successfully',
        token: token
    });
};

exports.logUserIn = async (req, res) => {
    const { email, password } = req.body;

    // Validate required fields
    if (!email || !password) {
        return res.status(400).json({ 
            error: 'Missing required fields: email, password' 
        });
    }
    // Check if user exists
    const user = await prisma.users.findUnique({
        where: {
            email: email
        }
    });
    if (!user) {
        return res.status(400).json({ 
            error: 'User not found with this email' 
        });
    }
    // Verify password
    const isPasswordValid = await bcrypt.compare(password, user.password);
    if (!isPasswordValid) {
        return res.status(400).json({ 
            error: 'Invalid password' 
        });
    }
    // create JWT token
    const token = generateToken({
        userId: user.id,
        name: user.name,
        email: user.email,
        role: user.role
    });

    res.status(200).json({
        message: 'Login successful',
        token: token
    });
}

exports.getAccountInfo = (req, res) => {
        res.status(200).json({
        message: 'User info retrieved successfully',
        user: req.userToken
    });
}

exports.deleteAccount = async (req, res) => {
    // Validate required fields
    if (!req.userToken) {
        return res.status(400).json({ 
            error: 'Token is missing' 
        });
    }
    const userId = req.userToken.userId;
    const token = req.jwt;

    try {
        // Execute a transaction to delete the user and add the token to the blacklist
        await prisma.$transaction([
            // Delete the user account
            prisma.users.delete({
                where: {
                    id: userId
                }
            }),
            // Add the token to the blacklist
            prisma.bljwts.create({
                data: {
                    jwt: token
                }
            }),
        ]);

        res.status(200).json({
            message: 'Account deleted and token blacklisted successfully'
        });
    } catch (error) {
        return res.status(500).json({
            error: 'Error deleting account or blacklisting token: ' + error.message
        });
    }
}

exports.sudoDeleteAccount = async (req, res) => {
    // Validate required fields
    if (!req.userToken || req.userToken.role !== 1) {
        return res.status(403).json({ 
            error: 'Only admin (role = 1) can delete accounts' 
        });
    }
    const accountToDelete = parseInt(req.params.id, 10); // Convert to integer (base 10)

    if (isNaN(accountToDelete)) {
        return res.status(500).json({
            error: 'Invalid account ID number'
        });
    }

    try {
        // Check if the user exists
        const user = await prisma.users.findUnique({
            where: {
                id: accountToDelete
            }
        });

        if (!user) {
            return res.status(500).json({
                error: 'User not found'
            });
        }
        // Delete the user account
        await prisma.users.delete({
            where: {
                id: accountToDelete
            }
        }),
        res.status(200).json({
            message: 'Account deleted identified by ID: ' + accountToDelete + ' successfully'
        });
    } catch (error) {
        return res.status(500).json({
            error: 'Error deleting account: ' + error.message
        });
    }
}

exports.sudoListAccount = async (req, res) => {
    // Validate required fields
    if (!req.userToken || req.userToken.role !== 1) {
        return res.status(403).json({ 
            error: 'Only admin (role = 1) can list accounts' 
        });
    }

    try {
        // List all user accounts
        const users = await prisma.users.findMany({
            select: {
                id: true,
                name: true,
                email: true,
                role: true
            }
        });

        res.status(200).json({
            users: users
        });
    } catch (error) {
        return res.status(500).json({
            error: 'Error listing accounts: ' + error.message
        });
    }
}

File: backend/routes/users/controller.py
from flask import request, jsonify, g
from ORM.db import get_db
from ORM.models import Users, blJWTs
from utils.jwt import generate_token
import bcrypt

def createNewAccount():
    data = request.get_json()
    name = data.get('name')
    email = data.get('email')
    password = data.get('password')

    if not name or not email or not password:
        return jsonify({'error': 'Missing required fields: name, email, password'}), 400

    with get_db() as db:
        existing_user = db.query(Users).filter(Users.email == email).count()
        if existing_user > 0:
            return jsonify({'error': 'User already exists with this email'}), 400

        hashed_password = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
        new_user = Users(name=name, email=email, password=hashed_password, role=0)
        db.add(new_user)
        db.commit()
        db.refresh(new_user)

        token = generate_token({
            'userId': new_user.id,
            'name': new_user.name,
            'email': new_user.email,
            'role': new_user.role
        })

    return jsonify({'message': 'Account created successfully', 'token': token}), 200

def logUserIn():
    data = request.get_json()
    email = data.get('email')
    password = data.get('password')

    if not email or not password:
        return jsonify({'error': 'Missing required fields: email, password'}), 400

    with get_db() as db:
        user = db.query(Users).filter(Users.email == email).first()
        if not user:
            return jsonify({'error': 'User not found with this email'}), 400

        is_password_valid = bcrypt.checkpw(password.encode('utf-8'), user.password.encode('utf-8'))
        if not is_password_valid:
            return jsonify({'error': 'Invalid password'}), 400

        token = generate_token({
            'userId': user.id,
            'name': user.name,
            'email': user.email,
            'role': user.role
        })

    return jsonify({'message': 'Login successful', 'token': token}), 200

def getAccountInfo():
    return jsonify({
        'message': 'User info retrieved successfully',
        'user': getattr(request, 'jwt_token', None)
    }), 200

def deleteAccount():
    if not hasattr(g, 'jwt_token'):
        return jsonify({'error': 'Token is missing'}), 400

    userId = g.jwt_token.get('userId')
    token = getattr(request, 'jwt', None)

    with get_db() as db:
        try:
            # Delete user and blacklist token in a transaction
            user = db.query(Users).filter(Users.id == userId).first()
            if user:
                db.delete(user)
            blacklisted = blJWTs(jwt=token)
            db.add(blacklisted)
            db.commit()
            return jsonify({'message': 'Account deleted and token blacklisted successfully'}), 200
        except Exception as error:
            db.rollback()
            return jsonify({'error': f'Error deleting account or blacklisting token: {str(error)}'}), 500

def sudoDeleteAccount(id):
    if not hasattr(g, 'jwt_token') or g.jwt_token.get('role') != 1:
        return jsonify({'error': 'Only admin (role = 1) can delete accounts'}), 403

    try:
        accountToDelete = int(id)
    except Exception:
        return jsonify({'error': 'Invalid account ID number'}), 500

    with get_db() as db:
        user = db.query(Users).filter(Users.id == accountToDelete).first()
        if not user:
            return jsonify({'error': 'User not found'}), 500

        try:
            db.delete(user)
            db.commit()
            return jsonify({'message': f'Account deleted identified by ID: {accountToDelete} successfully'}), 200
        except Exception as error:
            db.rollback()
            return jsonify({'error': f'Error deleting account: {str(error)}'}), 500

def sudoListAccount():
    if not hasattr(g, 'jwt_token') or g.jwt_token.get('role') != 1:
        return jsonify({'error': 'Only admin (role = 1) can list accounts'}), 403

    with get_db() as db:
        try:
            users = db.query(Users).with_entities(Users.id, Users.name, Users.email, Users.role).all()
            users_list = [
                {'id': u.id, 'name': u.name, 'email': u.email, 'role': u.role}
                for u in users
            ]
            return jsonify({'users': users_list}), 200
        except Exception as error:
            return jsonify({'error': f'Error listing accounts: {str(error)}'}), 500